### PR TITLE
Bonding on Android the Third

### PR DESCRIPTION
**Pull request by Tobi-Tobi**
As we're needing the bonding functionality and saw that there is help needed we decided to create our own pull request.

The error that [jeffhaynes](https://github.com/jefffhaynes) described in #724 also affects us => "bonding dialog appearing twice" also tracked by #497 

I've successfully tested that it doesn't occur anymore (Android 10 and 13). After testing around I found some bits missing from the #691 implementation. As #724 got written without the knowledge of #691 that is to be excepting.

So I sat down and merged the changes from #724 into #691.
 
The whole PR is tested with the following smartphones: 
1. Samsung S9 (Android 10)
2. Google Pixel 6A with latest Android 13

### CODE REVIEW

**Comment by janusw** [comment]
> As we're needing the bonding functionality and saw that there is help needed we decided to create our own pull request.

Thanks for working on this. Bonding seems to be our most-demanded feature right now 😃, so I think it would be very nice if we could get this into 3.0.

I'll try to do a basic review soon, and I also hope that @midoragh and @jefffhaynes will be able to test and review this (or anyone else interested in this feature).

Furthermore, we'll need to find out why the CI build on Windows broke. It does not seem that the changes in this PR are to blame. Instead it looks it the error appeared when the .NET SDK on the build host changed from 7.0.306 to 7.0.400.

**Comment by janusw** [comment]
> Furthermore, we'll need to find out why the CI build on Windows broke. It does not seem that the changes in this PR are to blame. Instead it looks it the error appeared when the .NET SDK on the build host changed from 7.0.306 to 7.0.400.

This is fixed via #737, but the CI build still fails here:

`BondStatusBroadcastReceiver.cs(37,33): error CS0117: 'OperatingSystem' does not contain a definition for 'IsAndroidVersionAtLeast' [D:\a\dotnet-bluetooth-le\dotnet-bluetooth-le\Source\Plugin.BLE\Plugin.BLE.csproj::TargetFramework=MonoAndroid10.0]`

@Tobi-Tobi Please fix!

**Comment by Tobi-Tobi** [comment]
@janusw Problem solved 👍

**Comment by janusw** [comment]
> @janusw Problem solved 👍

Alright, thanks for fixing!

Note that I will only merge this if @midoragh @jefffhaynes are happy with it. So guys, please speak up: Is this ok with you, or do you see any problems?

**Comment by jefffhaynes** [comment]
Sorry, I've been traveling for the last two weeks and don't have the device I was using with me but it looks correct to me. I'll try it as soon as I can.

File: Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs
Comment (janusw) [inline Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs:38]: One comment here: The usual pattern for this kind of SDK version check in other parts of the library is ...

```
#if NET6_0_OR_GREATER
                    OperatingSystem.IsAndroidVersionAtLeast(33)
#else
                    (Build.VERSION.SdkInt >= BuildVersionCodes.Tiramisu)
-#endif
```

Is there a particular reason why you chose to do it differently, or just "trying to make CI work somehow"? If I understand your comment correctly, then the `Tiramisu` version code is not available everywhere (although I don't fully understand why). Maybe just use the number instead, but stick to the rest of the pattern?

I think the idea of the above pattern was to use the modern way of doing this on recent .NET versions (i.e. `OperatingSystem.IsAndroidVersionAtLeast`), and the legacy way on .NET Standard / Xamarin (i.e. `Build.VERSION.SdkInt`).

I would prefer to make all such checks in a consistent manner, unless there is a good reason to deviate from it.

Comment (Tobi-Tobi) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs]: I changed it to your implementation, yet again that throws the error that Tiramisu is not available, see the failed github action:  
`D:\a\dotnet-bluetooth-le\dotnet-bluetooth-le\Source\Plugin.BLE\Android\BroadcastReceivers\BondStatusBroadcastReceiver.cs(37,59): error CS0117: 'BuildVersionCodes' does not contain a definition for 'Tiramisu'`  
Is a cast alright in this case? 

I didn't know that `OperatingSystem.IsAndroidVersionAtLeast(xyz)` is the preferred way in newer versions, updated it accordingly.

Comment (janusw) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs]: > I changed it to your implementation, yet again that throws the error that Tiramisu is not available (which I tried to evade by my implementation), see the failed github action: `D:\a\dotnet-bluetooth-le\dotnet-bluetooth-le\Source\Plugin.BLE\Android\BroadcastReceivers\BondStatusBroadcastReceiver.cs(37,59): error CS0117: 'BuildVersionCodes' does not contain a definition for 'Tiramisu'` Is a cast alright in this case?

I think the failure is due to the fact that we're not targeting Android 13 yet. So, yes, casting to int is ok with me. Thanks for your adjustments!

File: Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs
Comment (janusw) [inline Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs:44]: Another stupid question here: Why is the additional condition on NET7.0 necessary here? The variant with two arguments is only available on Android 33+ AFAICS, and you already have a check on the Android version above.

It is probably necessary because we are not targeting Android 13 in the classic Xamarin context, right? So maybe we should really update the targeted Android version for Xamarin. I can look into that.

Comment (janusw) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs:44]: > It is probably necessary because we are not targeting Android 13 in the classic Xamarin context, right? So maybe we should really update the targeted Android version for Xamarin. I can look into that.

see #740

Comment (Tobi-Tobi) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs]: image  
Older apps don't know the new overload added in api 33 right?

When using #740 changes and then removing the version check it runs into the following problem:
```
/Users/tobi/1_Code/3rd-Party/BT-Dotnet/dotnet-bluetooth-le-jeffhaynes/Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs(40,56): error CS1501: No overload for method 'GetParcelableExtra' takes 2 arguments [/Users/tobi/1_Code/3rd-Party/BT-Dotnet/dotnet-bluetooth-le-jeffhaynes/Source/Plugin.BLE/Plugin.BLE.csproj::TargetFramework=net6.0-android]
```
I'm not exactly sure if we can remove it even when using your changes. As the error may be a problem of my toolchain I pushed the MonoAndroid update with the removed check as it's own commit here. If the pipeline passes everything is fine, if not I'll revert it.

Comment (Tobi-Tobi) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs:44]: Yeah.. The CI also crashed because of it:  
```
Error: /Users/runner/work/dotnet-bluetooth-le/dotnet-bluetooth-le/Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs(40,56): 
error CS1501: No overload for method 'GetParcelableExtra' takes 2 arguments 
[/Users/runner/work/dotnet-bluetooth-le/dotnet-bluetooth-le/Source/Plugin.BLE/Plugin.BLE.csproj::TargetFramework=net6.0-android]
```

I reverted it. So the MonoAndroid update can be merged in its own branch.

Comment (janusw) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs]: Ok, very nice. Thanks.

File: Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs
Comment (janusw) [inline Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs:38]: With #740 merged, you should also be able to use the Tiramisu version code here (instead of casting to int). Then the comment in line 33 can go away.
Also: The whitespace/indentation in line 35 looks odd.

File: Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs
Comment (janusw) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs]: I re-triggered the CI and it succeeded now. 🎉

File: Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs
Comment (Tobi-Tobi) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs]: Ah, somehow a Tab sneaked in. I blame Rider 😄 

Changed the cast to BuildVersionCode.Tiramisu.   
Also fixed the indentation 🙈

Comment (janusw) [reply Source/Plugin.BLE/Android/BroadcastReceivers/BondStatusBroadcastReceiver.cs]: Awesome. I think this block of code is finally where it should be. Thanks 😁

**Comment by janusw** [comment]
So, originally I wanted to have a release 3.0 in August, now we're a bit late already. This PR is really the last bigger feature I'm willing to consider for the final release.

But since it combines elements of two earlier PRs for the same feature, I will not merge it unless the authors of the earlier PRs (@midoragh and @jefffhaynes) signal their approval. It hope that they could either review the code or do some testing (ideally both).

If we don't get any feedback by the end of September, I will publish the final release 3.0 without this feature (which will then have to wait for 3.1). I really hope we can still get it into 3.0, but I don't want to delay the release too much.

**Comment by jefffhaynes** [comment]
I'll test tonight. Thanks.

**Comment by jefffhaynes** [comment]
Initially had some super weird behavior but it seems like Android just didn't want to fully forget the previously bonded devices so I had to reset network settings (Android ftw).  However, after that everything worked smoothly.

Looks good, thanks for the update.

**Comment by janusw** [comment]
> Looks good, thanks for the update.

Thanks for the positive feedback, @jefffhaynes !

@midoragh Do you think you'll be able to do some code review and/or testing in the next two weeks?

**Review by janusw — APPROVED** [review]

**Comment by janusw** [comment]
It looks like we're not getting any comment from @midoragh here, but we have an approval from @jefffhaynes and all my comments have been addressed, so I will simply merge this now (and tag an rc for 3.0). Thanks for the contribution, @Tobi-Tobi !

**Comment by janusw** [comment]
I prepared the tag and uplodaded to nuget. If any problems come up (or we get a late review), we can still fix it for the final version 3.0, which I would like to release in about a week from now (if no blockers appear in the meantime).

**Comment by Tobi-Tobi** [comment]
> I prepared the tag and uplodaded to nuget. If any problems come up (or we get a late review), we can still fix it for the final version 3.0, which I would like to release in about a week from now (if no blockers appear in the meantime).

Thank you @janusw  
We're also going to update to the RC soon (early next week). If any problems arise I'll create an issue or fix it. 👍

**Comment by Tobi-Tobi** [comment]
Update: 

We tested it some minutes ago, still works as expected with the RC. 👍

**Pull request merged**
